docs/hdl_coding_guideline.md

```diff
@@ -0,0 +1,710 @@
+  * *Should* rules are advisory rules. They suggest the recommended way of doing 
+things.
+  * *Must* rules are mandatory requirements.
+
+The coding rules are intended to be applied to HDL modules written using VHDL 
+or Verilog.
+
+When an external IP is used, the naming conventions practiced by the IP *should* be kept, 
+even if they do not match the rules specified in this document.
+
+##  Coding Style
+
+### A. Layout
+
+**A1**: For indentation, spaces **must** be used instead of tabs. This allows 
+the code to be properly visualized by any editor. **Do not** leave spaces at 
+the end of a line. The following editor settings **must** be used: 
+*Tab Size*: 2, *Indent Size*: 2
+
+
+**A2**: One white space **must** be inserted around operators, such as 
+=, ==, &&, ||, &, |, ^, etc.
+
+#### Example:
+
+##### Incorrect:
+```verilog
+if((my_signal1==1'b0)&&(my_bus[3:0]==4'd5))
+```
+
+##### Correct:
+```verilog
+if ((my_signal == 1'b0) && (my_bus[3:0] == 4'd5))
+```
+
+
+**A3**: The _always_ block *should* have a space before **@** symbol.
+
+#### Example:
+
+##### Incorrect:
+```verilog
+always@(posedge clk) begin
+  ...
+end
+```
+##### Correct:
+```verilog
+always @(posedge clk) begin
+  ...
+end
+```
+
+
+**A4**: The Verilog ``begin``/``end`` block **must** always be used, even if there is 
+only one statement. This makes adding lines of code much easier and with fewer errors.
+
+
+**A5**: Indentation levels **must** be used to show code nesting. Blank lines 
+may be used as desired to improve code readability, but _not_ in all cases.
+
+#### Example:
+
+##### Incorrect:
+```verilog
+if (my_signal == 1'b0) begin
+  if (my_bus[3:0]==4'd5) begin
+  statement1;
+  statement2;
+  end
+statement3;
+statement4;
+end
+else
+statement5;
+```
+
+##### Correct:
+```verilog
+if (my_signal == 1'b0) begin
+  if (my_bus[3:0] == 4'd5) begin
+    statement1;
+    statement2;
+  end
+  statement3;
+  statement4;
+end else begin
+  statement5;
+end
+```
+
+
+**A6**: In a ``case`` definition, indentation levels **must** be used to offset 
+the statements that are encapsulated, but the use of blank lines can be used or 
+omitted to best show the statement groupings (if really necessary). ``end`` should 
+be indented as in the correct example.
+
+#### Example:
+
+##### Incorrect:
+```verilog
+case ( my_bus[3:0] )
+  4'b0000 : my_signal1 = TRUE;
+  4'b0001 : my_signal1 = FALSE;
+  4'b0010 :
+  begin
+  my_signal1 = TRUE;
+  my_signal2 = FALSE;
+  end
+  4'b0100 : my_signal2 = FALSE;
+  default : my_signal1 = TRUE;
+endcase
+```
+
+##### Correct:
+```verilog
+case (my_bus[3:0])
+  4'b0000: begin
+    my_signal1 = TRUE;
+    end
+  4'b0001: begin
+    my_signal1 = FALSE;
+    end
+  4'b0010: begin
+    my_signal1 = TRUE;
+    my_signal2 = FALSE;
+    end
+  4'b0100: begin
+    my_signal2 = FALSE;
+    end
+  default: begin
+    my_signal1 = TRUE;
+    end
+endcase
+```
+
+
+**A7**: Alignment **must** be used in declarations, assignments, multi-line 
```

Comment:
We have a lot of generic files, in which some bus widths are dependent on more than one parameter. If we got with a "must", than we have to pass the calculations through a new parameter. I wold keep this at a "should", for the time being, considering the alignment rule is broken in some files.